### DISCUSSION TITLE
How do I get query data from a database function?

### DISCUSSION THREAD

**Discussion by Rayyanmaq1** [supabase/supabase#2741] [Questions] [answered]
Having issue in return type in rcp() method if return type is record it give following error
"a column definition list is required for functions returning "record" sql" 
and other return type such as json and jsonb cann't be selected

<img width="642" alt="Screenshot 2021-08-06 at 12 03 50 PM" src="https://user-images.githubusercontent.com/71687063/128470703-184df628-1cfe-4113-a090-3d57ec31aa1b.png">

**Comment by silentworks** [comment, marked as answer]
You need to tell the function what to return. Change the body of the function to the below:

```sql
BEGIN
  RETURN QUERY 
    SELECT post_id, user_id FROM posts 
    WHERE user_id IN(SELECT user_to FROM follow WHERE user_from = currentuseruid) AND privacy = 'My Friends';
END;
```

**Reply by Rayyanmaq1** [reply]
<img width="355" alt="Screenshot 2021-08-06 at 12 29 50 PM" src="https://user-images.githubusercontent.com/71687063/128473713-1636201f-6550-4498-aebc-4807b3b0e27d.png">

**Reply by Rayyanmaq1** [reply]
about error if i use postgresql but when i use sql them it gives Error: syntax error at or near "RETURN"

**Reply by silentworks** [reply]
Just use the SQL editor and run the following SQL, you might need to modify the `post_id` type as I'm not sure if you are using `integer` or `uuid` for it.

```sql
CREATE OR REPLACE FUNCTION public.getUserPosts(currentuseruid uuid) RETURNS TABLE(post_id integer, user_id uuid) AS $$
    BEGIN
        RETURN QUERY
            SELECT post_id, user_id 
            FROM posts 
            WHERE user_id IN(SELECT user_to FROM follow WHERE user_from = currentuseruid) AND privacy = 'My Friends';
    END;
$$ LANGUAGE plpgsql;
```

**Reply by Rayyanmaq1** [reply]
Its working thanks aalot
but what if i have to return every column of the table

**Reply by silentworks** [reply]
You will have to list all the columns you want to return. Otherwise, you have to create a view and use that with RETURNS instead.

Also please change this post title to **How do I return query data from a database function?**

**Reply by awaaisali** [reply]
@silentworks Can you please guide me on how to invoke that function?

**Reply by b2m9** [reply]
@awaaisali Here are the docs, they walk you through: https://supabase.com/docs/guides/database/functions

**Reply by awaaisali** [reply]
@b2m9 Thanks
